rest_framework/decorators.py

```diff
@@ -46,11 +47,20 @@ def decorator(func):
         allowed_methods = set(http_method_names) | {'options'}
         WrappedAPIView.http_method_names = [method.lower() for method in allowed_methods]
 
-        def handler(self, *args, **kwargs):
+        def sync_handler(self, *args, **kwargs):
             return func(*args, **kwargs)
 
-        for method in http_method_names:
-            setattr(WrappedAPIView, method.lower(), handler)
+        async def async_handler(self, *args, **kwargs):
+            return await func(*args, **kwargs)
+
+        view_is_async = asyncio.iscoroutinefunction(func)
+
+        if view_is_async:
```

Comment:
You can move the definitions of `sync_handler` and `async_handler` into the if statements below

Reply:
I disagree. Redefining a function on each execution will likely have a higher performance impact than the current implementation.

Plus would impact readability.

Reply:
I'd argue the opposite.

1. It improves performance. Currently, both `sync` and `async` handlers are defined per definition of `decorator` per call of `api_view` (one is used and the other is discarded). Moving definitions into their respective branches would remove the wasted definition.
2. It improves readability. Because the change moves the relevant code to the _only_ place it is used -- inside their respective branches.

Reply:
@Archmonger I have made the change according to @dongyuzheng suggestion, I hope you will like the idea.

Reply:
Looks good 👍